File: src/main/java/com/upokecenter/mail/Address.java
package com.upokecenter.mail;
/*
Written by Peter O. in 2014.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://peteroupc.github.io/
 */

import java.util.*;

import com.upokecenter.util.*;
import com.upokecenter.text.*;

    /**
     * Represents an email address.
     */
  public class Address {
    private final String localPart;

    /**
     * Determines whether this object and another object are equal.
     * @param obj The parameter {@code obj} is an arbitrary object.
     * @return {@code true} if this object and another object are equal; otherwise,
     * {@code false}.
     */
    @Override public boolean equals(Object obj) {
      Address other = ((obj instanceof Address) ? (Address)obj : null);
      return other != null && this.localPart.equals(other.localPart,
  StringComparison.Ordinal) &&
        this.domain.equals(other.domain, StringComparison.Ordinal);
    }

    /**
     * Gets the local part of this email address (the part before the "@" sign).
     * @return The local part of this email address (the part before the "@" sign).
     */
    public final String getLocalPart() {
        return this.localPart;
      }
    private static String DomainToString(String domain, boolean useALabelDomain) {
      String dom = domain;
      if (useALabelDomain && dom.length() > 0 && dom.charAt(0) != '[') {
        dom = Idna.EncodeDomainName(domain);
      }
      return dom;
    }
    static String LocalPartToString(String localPart) {
      if (localPart.length() > 0 && HeaderParser.ParseDotAtomText(
        localPart,
        0,
        localPart.length(),
        null) == localPart.length()) {
        return localPart;
      } else {
        StringBuilder sb = new StringBuilder();
        sb.append('"');
        for (int i = 0; i < localPart.length(); ++i) {
          char c = localPart.charAt(i);
          if (c == 0x20 || c == 0x09) {
            sb.append(c);
          } else if (c == '"' || c == 0x7f || c == '\\' || c < 0x20) {
            sb.append('\\');
            sb.append(c);
          } else {
            sb.append(c);
          }
        }
        sb.append('"');
        return sb.toString();
      }
    }

    void AppendThisAddress(HeaderEncoder encoder) {
      String lp = LocalPartToString(this.localPart);
      String domainstr = DomainToString(this.domain, true);
      int length = DataUtilities.CodePointLength(lp);
      int length2 = DataUtilities.CodePointLength(domainstr);
      if (length2 + length + 1 <= Message.MaxRecHeaderLineLength - 1) {
        // Avoid breaking email addresses if it can comfortably
        // fit the recommended line length
        int tlength = (int)(length2 + length + 1);
        encoder.AppendSymbolWithLength(lp + "@" + domainstr, tlength);
      } else {
        encoder.AppendSymbolWithLength(lp, length);
        encoder.AppendSymbol("@");
        encoder.AppendSymbolWithLength(domainstr, length);
      }
    }

    /**
     * Converts this address object to a text string.
     * @return A string representation of this object.
     */
    @Override public String toString() {
      HeaderEncoder sa = new HeaderEncoder(Message.MaxRecHeaderLineLength, 15);
      this.AppendThisAddress(sa);
      return sa.toString();
    }

    private boolean IsTooLong() {
      String lp = LocalPartToString(this.localPart);
      String domainstr = DomainToString(this.domain, true);
      String domain2 = DomainToString(this.domain, false);
      // Maximum OCTET length per line for an Internet message minus 1;
      // we check if the length exceeds that number (thus excluding the space
      // character of a folded line).
      if (DataUtilities.GetUtf8Length(lp, true) >
    Message.MaxHardHeaderLineLength - 1) {
        return true;
      }
      if (DataUtilities.GetUtf8Length(domainstr, true) >
        Message.MaxHardHeaderLineLength - 1) {
        return true;
      }
      return (DataUtilities.GetUtf8Length(domain2, true) >
        Message.MaxHardHeaderLineLength - 1) ? true : false;
    }

    /**
     * Returns a hash code for this address object. No application or process
     * identifiers are used in the hash code calculation.
     * @return A hash code for this instance.
     */
    @Override public int hashCode() {
      int valueHashCode = -1524613162;
      if (this.domain != null) {
        for (int i = 0; i < this.domain.length(); ++i) {
          valueHashCode *= -1521134295 + this.domain.charAt(i);
        }
      }
      if (this.localPart != null) {
        for (int i = 0; i < this.localPart.length(); ++i) {
          valueHashCode *= -1521134295 + this.localPart.charAt(i);
        }
      }
      return valueHashCode;
    }

    private final String domain;

    /**
     * Gets the domain of this email address (the part after the "@" sign).
     * @return The domain of this email address (the part after the "@" sign).
     */
    public final String getDomain() {
        return this.domain;
      }

    /**
     * Initializes a new instance of the {@link Address} class.
     * @param addressValue The parameter {@code addressValue} is a text string.
     * @throws NullPointerException The parameter {@code addressValue} is null.
     * @throws IllegalArgumentException AddressValue is empty.; Address doesn't contain a
     * '@' sign; Invalid local part; Expected '@' sign after local part;
     * Expected domain after '@'; Invalid domain; Address too long.
     */
    public Address(String addressValue) {
      if (addressValue == null) {
        throw new NullPointerException("addressValue");
      }
      if (addressValue.length() == 0) {
        throw new IllegalArgumentException("addressValue is empty.");
      }
      if (addressValue.indexOf('@') < 0) {
        throw new IllegalArgumentException("Address doesn't contain a '@' sign");
      }
      int localPartEnd = HeaderParser.ParseLocalPartNoCfws(
        addressValue,
        0,
        addressValue.length(),
        null);
      if (localPartEnd == 0) {
        throw new IllegalArgumentException("Invalid local part");
      }
      if (localPartEnd >= addressValue.length() ||
     addressValue.charAt(localPartEnd) != '@') {
        throw new IllegalArgumentException("Expected '@' sign after local part");
      }
      if (localPartEnd + 1 == addressValue.length()) {
        throw new IllegalArgumentException("Expected domain after '@'");
      }
      int domainEnd = HeaderParser.ParseDomainNoCfws(
        addressValue,
        localPartEnd + 1,
        addressValue.length(),
        null);
      if (domainEnd != addressValue.length()) {
        throw new IllegalArgumentException("Invalid domain");
      }
      this.localPart = HeaderParserUtility.ParseLocalPart(
        addressValue,
        0,
        localPartEnd);
      this.domain = HeaderParserUtility.ParseDomain(
        addressValue,
        localPartEnd + 1,
        addressValue.length());
      // Check length restrictions.
      if (this.IsTooLong()) {
        throw new IllegalArgumentException("Address too long");
      }
    }

    Address(String localPart, String domain) {
      if (localPart == null) {
        throw new NullPointerException("localPart");
      }
      if (domain == null) {
        throw new NullPointerException("domain");
      }
      this.localPart = localPart;
      this.domain = domain;
      // Check length restrictions.
      if (this.IsTooLong()) {
        throw new IllegalArgumentException("Address too long");
      }
    }
  }

File: MailLib/PeterO/Mail/Address.cs
/*
Written by Peter O. in 2014.
Any copyright is dedicated to the Public Domain.
http://creativecommons.org/publicdomain/zero/1.0/
If you like this, you should donate to Peter O.
at: http://peteroupc.github.io/
 */
using System;
using System.Collections.Generic;
using System.Text;
using PeterO;
using PeterO.Text;

namespace PeterO.Mail {
    /// <summary>Represents an email address.</summary>
  public class Address {
    private readonly string localPart;

    /// <summary>Determines whether this object and another object are
    /// equal.</summary>
    /// <param name='obj'>The parameter <paramref name='obj'/> is an
    /// arbitrary object.</param>
    /// <returns><c>true</c> if this object and another object are equal;
    /// otherwise, <c>false</c>.</returns>
    public override bool Equals(object obj) {
      var other = obj as Address;
      return other != null && this.localPart.Equals(other.localPart,
  StringComparison.Ordinal) &&
        this.domain.Equals(other.domain, StringComparison.Ordinal);
    }

    /// <summary>Gets the local part of this email address (the part before
    /// the "@" sign).</summary>
    /// <value>The local part of this email address (the part before the
    /// "@" sign).</value>
    public string LocalPart {
      get {
        return this.localPart;
      }
    }
    private static string DomainToString(string domain, bool useALabelDomain) {
      string dom = domain;
      if (useALabelDomain && dom.Length > 0 && dom[0] != '[') {
        dom = Idna.EncodeDomainName(domain);
      }
      return dom;
    }
    internal static string LocalPartToString(string localPart) {
      if (localPart.Length > 0 && HeaderParser.ParseDotAtomText(
        localPart,
        0,
        localPart.Length,
        null) == localPart.Length) {
        return localPart;
      } else {
        var sb = new StringBuilder();
        sb.Append('"');
        for (int i = 0; i < localPart.Length; ++i) {
          char c = localPart[i];
          if (c == 0x20 || c == 0x09) {
            sb.Append(c);
          } else if (c == '"' || c == 0x7f || c == '\\' || c < 0x20) {
            sb.Append('\\');
            sb.Append(c);
          } else {
            sb.Append(c);
          }
        }
        sb.Append('"');
        return sb.ToString();
      }
    }

    internal void AppendThisAddress(HeaderEncoder encoder) {
      string lp = LocalPartToString(this.localPart);
      string domainstr = DomainToString(this.domain, true);
      int length = DataUtilities.CodePointLength(lp);
      int length2 = DataUtilities.CodePointLength(domainstr);
      if (length2 + length + 1 <= Message.MaxRecHeaderLineLength - 1) {
        // Avoid breaking email addresses if it can comfortably
        // fit the recommended line length
        var tlength = (int)(length2 + length + 1);
        encoder.AppendSymbolWithLength(lp + "@" + domainstr, tlength);
      } else {
        encoder.AppendSymbolWithLength(lp, length);
        encoder.AppendSymbol("@");
        encoder.AppendSymbolWithLength(domainstr, length);
      }
    }

    /// <summary>Converts this address object to a text string.</summary>
    /// <returns>A string representation of this object.</returns>
    public override string ToString() {
      var sa = new HeaderEncoder(Message.MaxRecHeaderLineLength, 15);
      this.AppendThisAddress(sa);
      return sa.ToString();
    }

    private bool IsTooLong() {
      string lp = LocalPartToString(this.localPart);
      string domainstr = DomainToString(this.domain, true);
      string domain2 = DomainToString(this.domain, false);
      // Maximum OCTET length per line for an Internet message minus 1;
      // we check if the length exceeds that number (thus excluding the space
      // character of a folded line).
      if (DataUtilities.GetUtf8Length(lp, true) >
    Message.MaxHardHeaderLineLength - 1) {
        return true;
      }
      if (DataUtilities.GetUtf8Length(domainstr, true) >
        Message.MaxHardHeaderLineLength - 1) {
        return true;
      }
      return (DataUtilities.GetUtf8Length(domain2, true) >
        Message.MaxHardHeaderLineLength - 1) ? true : false;
    }

    /// <summary>Returns a hash code for this address object. No
    /// application or process identifiers are used in the hash code
    /// calculation.</summary>
    /// <returns>A hash code for this instance.</returns>
    public override int GetHashCode() {
      var hashCode = -1524613162;
      if (this.domain != null) {
        for (var i = 0; i < this.domain.Length; ++i) {
          hashCode *= -1521134295 + this.domain[i];
        }
      }
      if (this.localPart != null) {
        for (var i = 0; i < this.localPart.Length; ++i) {
          hashCode *= -1521134295 + this.localPart[i];
        }
      }
      return hashCode;
    }

    private readonly string domain;

    /// <summary>Gets the domain of this email address (the part after the
    /// "@" sign).</summary>
    /// <value>The domain of this email address (the part after the "@"
    /// sign).</value>
    public string Domain {
      get {
        return this.domain;
      }
    }

    /// <summary>Initializes a new instance of the
    /// <see cref='PeterO.Mail.Address'/> class.</summary>
    /// <param name='addressValue'>The parameter <paramref name='addressValue'/> is a text string.</param>
    /// <exception cref='System.ArgumentNullException'>The parameter <paramref name='addressValue'/> is null.</exception>
    /// <exception cref='System.ArgumentException'>AddressValue is empty.; Address
    /// doesn't contain a '@' sign; Invalid local part; Expected '@' sign
    /// after local part; Expected domain after '@'; Invalid domain;
    /// Address too long.</exception>
    public Address(string addressValue) {
      if (addressValue == null) {
        throw new ArgumentNullException(nameof(addressValue));
      }
      if (addressValue.Length == 0) {
        throw new ArgumentException("addressValue is empty.");
      }
      if (addressValue.IndexOf('@') < 0) {
        throw new ArgumentException("Address doesn't contain a '@' sign");
      }
      int localPartEnd = HeaderParser.ParseLocalPartNoCfws(
        addressValue,
        0,
        addressValue.Length,
        null);
      if (localPartEnd == 0) {
        throw new ArgumentException("Invalid local part");
      }
      if (localPartEnd >= addressValue.Length ||
     addressValue[localPartEnd] != '@') {
        throw new ArgumentException("Expected '@' sign after local part");
      }
      if (localPartEnd + 1 == addressValue.Length) {
        throw new ArgumentException("Expected domain after '@'");
      }
      int domainEnd = HeaderParser.ParseDomainNoCfws(
        addressValue,
        localPartEnd + 1,
        addressValue.Length,
        null);
      if (domainEnd != addressValue.Length) {
        throw new ArgumentException("Invalid domain");
      }
      this.localPart = HeaderParserUtility.ParseLocalPart(
        addressValue,
        0,
        localPartEnd);
      this.domain = HeaderParserUtility.ParseDomain(
        addressValue,
        localPartEnd + 1,
        addressValue.Length);
      // Check length restrictions.
      if (this.IsTooLong()) {
        throw new ArgumentException("Address too long");
      }
    }

    internal Address(string localPart, string domain) {
      if (localPart == null) {
        throw new ArgumentNullException(nameof(localPart));
      }
      if (domain == null) {
        throw new ArgumentNullException(nameof(domain));
      }
      this.localPart = localPart;
      this.domain = domain;
      // Check length restrictions.
      if (this.IsTooLong()) {
        throw new ArgumentException("Address too long");
      }
    }
  }
}
